src/BPGates.jl

```diff
@@ -518,6 +534,80 @@ function QuantumClifford.apply!(state::BellState, g::PauliNoiseOp)
     return state
 end
 
+# Type for abstracting away noisy operations on bell states.
+abstract type AbstractNoiseBellOp <: BellOp end
+
+"""
+PauliZOp(idx, pz) causes qubit at idx `idx` to have a Z flip with probabilty `pz`.
+"""
+struct PauliZOp <: AbstractNoiseBellOp
+    idx::Int
+    # assert that 0 <= pz <= 1? not strictly necessary. it wouldn't make sense, but it
+    # also wouldn't break the code. natural floor and ceiling here.
+    pz::Float64
+end
+
+function QuantumClifford.apply!(state::BellState, g::PauliZOp)
+    i = g.idx
+    r = rand()
+    if r<g.pz
+        # I believe this actually is a Z flip.
+        apply!(state, BellPauliPermutation(2, i))
+    end
+    return state
+end
+
+# TODO: continue from here
+# TODO: assert that the values add to 1
+# This is NOT at all 'extensible' (though, it is 'fast')
+# TODO: Honestly this doesn't make that much sense to me. Think about why
+# this works.
+# Defines the probabilities of transitioning from one bell state to another
+function get_mixed_transition_probs(λ::Float64, cur_state_idx::Int)
+    # 0 <= λ <= 1 (λ = 1 - e ^(-t / T1))
+    mixed_state_tuple = (
+        (0.5 * λ^2 - λ + 1, 0.5 * λ * (1 - λ), 0.5 * λ^2, 0.5 * λ * (1 - λ)),
+        (0.5 * λ, 1 - λ, 0.5 * λ, 0),
+        (0.5 * λ^2, 0.5 * λ * (1 - λ), 0.5 * λ^2 - λ + 1, 0.5 * λ * (1 - λ)),
+        (0.5 * λ, 0, 0.5 * λ, 1 - λ)
+    )
+    return mixed_state_tuple[cur_state_idx]
+end
+
+"""
+MixedStateOp(idx, lambda) causes Bell state at idx `idx` to change to another
+Bell state determined by `mixed_state_tuple` and `lambda`.
+"""
+struct MixedStateOp <: AbstractNoiseBellOp
```

Comment:
Why is this called `MixedStateOp`? Semantically I am not sure I am following the naming convention.

Reply:
Mainly because in the original BPGates.jl, even probabilistic applications are suffixed by "Op".

Maybe it makes more sense to say that this is specifically a noise operation, so calling this "AmplitudeDampingTwirlingOp" makes more sense?